cmd/fil-daggregator/main.go

```diff
@@ -0,0 +1,345 @@
+	return <-errCh
+}
+
+// pulls cids from an AllKeysChan and sends them concurrently via multiple workers to an API
+func writeoutBlocks(externalCtx context.Context, opts *opts, bs blockstore.Blockstore) error {
+
+	ctx, shutdownWorkers := context.WithCancel(externalCtx)
+	defer shutdownWorkers()
+
+	akc, err := bs.AllKeysChan(ctx)
+	if err != nil {
+		return err
+	}
+
+	var lastPct uint64
+	blocksDone := new(uint64)
+
+	// this works because of how AllKeysChan behaves on rambs
+	blocksTotal := uint64(len(akc))
+
+	maxWorkers := opts.IpfsAPIMaxWorkers
+	errCh := make(chan error, 1+maxWorkers) // our own error plus one from each worker
+
+	// watchdog for progress and to bail early if an error appears
+	go func() {
+		var progressTick <-chan time.Time
+		if opts.ShowProgress {
+			fmt.Fprint(os.Stderr, "0% of blocks written\r")
+			t := time.NewTicker(250 * time.Millisecond)
+			progressTick = t.C
+			defer t.Stop()
+		}
+		for {
+			select {
+			case <-ctx.Done():
+				errCh <- ctx.Err()
+				return
+			case err := <-errCh:
+				errCh <- err // put error back where we got it from
+				if err != nil {
+					shutdownWorkers()
+				}
+				return
+			case <-progressTick:
+				curPct := 100 * atomic.LoadUint64(blocksDone) / blocksTotal
+				if curPct != lastPct {
+					lastPct = curPct
+					fmt.Fprintf(os.Stderr, "%d%% of blocks written\r", lastPct)
+				}
+			}
+		}
+	}()
+
+	var wg sync.WaitGroup
+
+	for maxWorkers > 0 {
+		maxWorkers--
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			api := ipfsapi.NewShell(opts.IpfsAPI)
+			api.SetTimeout(time.Second * time.Duration(opts.IpfsAPITimeoutSecs))
+
+			for {
+				select {
+
+				case <-ctx.Done():
+					return
+
+				case c, chanOpen := <-akc:
+
+					if !chanOpen {
+						return
+					}
+
+					blk, err := bs.Get(c)
+					if err != nil {
+						errCh <- err
+						return
+					}
+
+					// copied entirety of ipfsapi.BlockPut() to be able to pass in our own ctx 🤮
+					res := new(struct{ Key string })
+					err = api.Request("block/put").
+						Option("format", cid.CodecToStr[c.Prefix().Codec]).
+						Option("mhtype", multihash.Codes[c.Prefix().MhType]).
+						Option("mhlen", c.Prefix().MhLength).
+						Body(
+							ipfsfiles.NewMultiFileReader(
+								ipfsfiles.NewSliceDirectory([]ipfsfiles.DirEntry{
+									ipfsfiles.FileEntry(
+										"",
+										ipfsfiles.NewBytesFile(blk.RawData()),
+									),
+								}),
+								true,
+							),
+						).
+						Exec(ctx, res)
+					// end of 🤮
+
+					if err != nil {
+						errCh <- err
+						return
+					}
+
+					if res.Key != c.String() {
+						errCh <- xerrors.Errorf("unexpected cid mismatch after /block/put: expected %s but got %s", c, res.Key)
+						return
+					}
+
+					if opts.ShowProgress {
+						atomic.AddUint64(blocksDone, 1)
+					}
+				}
+			}
+		}()
+	}
+
+	wg.Wait()
+	shutdownWorkers()
+
+	// feeder drain if anything remains
+	for len(akc) > 0 {
+		<-akc
+	}
+
+	errCh <- externalCtx.Err()
+	return <-errCh
```

Comment:
hmm, why would you send to a channel and then immediately after receive from it?

Reply:
Higher up: https://github.com/filecoin-project/go-dagaggregator-unixfs/pull/4/files#diff-a4e4a43b6b5adae08f8c7657c39f65305a1c036855ee5049c792a308c7dc2f84R140

I want to get the "first error whatever it was". Open to learn what's a cleaner way to do this when we chat